lib/govuk_navigation_helpers/grouped_related_links.rb

```diff
@@ -0,0 +1,48 @@
+module GovukNavigationHelpers
+  # Take a content item and group the related links according to an algorithm
+  # that is intended to display the related links into three groups, depending
+  # on how much they have in common with the main content item.
+  #
+  # @private
+  class GroupedRelatedLinks
+    attr_reader :content_item,
+
+      # This will contain related items that have the same `parent` (breadcrumb)
+      # as the content item.
+      :related_with_parent_in_common,
+
+      # This will contain related items that have a grandparent in common with
+      # the content item.
+      :related_with_grandparent_in_common,
+
+      # This will contain the related items that have a completely different
+      # parent/breadcrumb from the content item.
+      :related_with_nothing_in_common
+
+    def initialize(content_item)
+      @content_item = content_item
+      @related_with_parent_in_common = []
+      @related_with_grandparent_in_common = []
+      @related_with_nothing_in_common = []
+      group_related_links!
+    end
+
+  private
+
+    def group_related_links!
+      content_item.related_links.each do |related_item|
+        if content_item.parent && related_item.parent
```

Comment:
What do you think of refactoring all these `if`, `else`, `elsif` clauses? I personally find it a bit hard to follow. It could be a bit more verbose, something like this:

```
def related_with_parent_in_common
  @related_with_parent_in_common ||=
    content_item.related_links.select { |related_item| related_parent?(related_item) }
end

def related_with_grandparent_in_common
  @related_with_grandparent_in_common ||=
    content_item.related_links.select { |related_item| related_grandparent?(related_item) }
end

def related_with_nothing_in_common
  @related_with_nothing_in_common ||=
    content_item.related_links.select { |related_item| unrelated?(related_item) }
end

private

def related_parent?(related_item)
  ...
end

...
```

What do you think?